flaky test: TestE2ENullAlterWithReplicas

```
2026-05-01T12:56:33.9535977Z test-1           | --- FAIL: TestE2ENullAlterWithReplicas (1.52s)
2026-05-01T12:56:33.9536073Z test-1           |     migration_test.go:109: 
2026-05-01T12:56:33.9536239Z test-1           |         	Error Trace:	/app/pkg/migration/migration_test.go:109
2026-05-01T12:56:33.9536366Z test-1           |         	Error:      	Received unexpected error:
2026-05-01T12:56:33.9536652Z test-1           |         	            	replica is not healthy: IO thread not running on 45c4c2106299:3306 (Replica_IO_State: )
2026-05-01T12:56:33.9536796Z test-1           |         	Test:       	TestE2ENullAlterWithReplicas

```

This likely has something to do with the test setup / test-only failure.

--- a/compose/replication-tls/init-ci-replica.sql
+++ b/compose/replication-tls/init-ci-replica.sql
@@ -1,1 +1,2 @@
 change replication source to source_host='mysql',source_user='rsandbox',source_password='rsandbox',source_auto_position=1,get_source_public_key=1;
+start replica;

--- a/pkg/migration/migration_test.go
+++ b/pkg/migration/migration_test.go
@@ -99,6 +99,7 @@ func TestE2ENullAlterWithReplicas(t *testing.T) {
 	if replicaDSN == "" {
 		t.Skip("skipping replica tests because REPLICA_DSN not set")
 	}
+	testutils.WaitForReplicaHealthy(t, replicaDSN, 30*time.Second)
 	testutils.NewTestTable(t, "replicatest", `CREATE TABLE replicatest (
 		id int(11) NOT NULL AUTO_INCREMENT,
 		name varchar(255) NOT NULL,

--- a/pkg/testutils/testing.go
+++ b/pkg/testutils/testing.go
@@ -11,6 +11,7 @@ import (
 	"sync"
 	"sync/atomic"
 	"testing"
+	"time"
 
 	"github.com/go-sql-driver/mysql"
 	"github.com/stretchr/testify/assert"
@@ -149,6 +150,74 @@ func IsMinimalRBRTestRunner(t *testing.T) bool {
 		}
 	})
 	return isRBRTestRunnerCached
+}
+
+// WaitForReplicaHealthy polls SHOW REPLICA STATUS until both the IO and SQL
+// threads report Yes, or the timeout elapses. On timeout it skips the test
+// with an infra-attribution message — when the replica isn't running at the
+// start of a test, that's almost always a CI setup issue, not a Spirit bug,
+// so failing late inside m.Run() is misleading.
+func WaitForReplicaHealthy(t *testing.T, dsn string, timeout time.Duration) {
+	t.Helper()
+	db, err := sql.Open("mysql", dsn)
+	require.NoError(t, err)
+	defer func() { _ = db.Close() }()
+
+	deadline := time.Now().Add(timeout)
+	var lastIO, lastSQL, lastIOState string
+	for {
+		ioRunning, sqlRunning, ioState, err := readReplicaStatus(t.Context(), db)
+		if err == nil {
+			if ioRunning == "Yes" && sqlRunning == "Yes" {
+				return
+			}
+			lastIO, lastSQL, lastIOState = ioRunning, sqlRunning, ioState
+		}
+		if time.Now().After(deadline) {
+			t.Skipf("test infra: replica not healthy after %s "+
+				"(Replica_IO_Running=%q Replica_SQL_Running=%q Replica_IO_State=%q lastErr=%v); "+
+				"this indicates a CI setup issue, not a Spirit bug",
+				timeout, lastIO, lastSQL, lastIOState, err)
+		}
+		time.Sleep(200 * time.Millisecond)
+	}
+}
+
+func readReplicaStatus(ctx context.Context, db *sql.DB) (ioRunning, sqlRunning, ioState string, err error) {
+	rows, err := db.QueryContext(ctx, "SHOW REPLICA STATUS")
+	if err != nil {
+		return "", "", "", err
+	}
+	defer func() { _ = rows.Close() }()
+	cols, err := rows.Columns()
+	if err != nil {
+		return "", "", "", err
+	}
+	if !rows.Next() {
+		if err := rows.Err(); err != nil {
+			return "", "", "", err
+		}
+		return "", "", "", fmt.Errorf("SHOW REPLICA STATUS returned no rows")
+	}
+	values := make([]any, len(cols))
+	for i := range values {
+		values[i] = new(sql.NullString)
+	}
+	if err := rows.Scan(values...); err != nil {
+		return "", "", "", err
+	}
+	for i, name := range cols {
+		v := values[i].(*sql.NullString).String
+		switch name {
+		case "Replica_IO_Running":
+			ioRunning = v
+		case "Replica_SQL_Running":
+			sqlRunning = v
+		case "Replica_IO_State":
+			ioState = v
+		}
+	}
+	return ioRunning, sqlRunning, ioState, nil
 }
 
 // EvenOddHasher is a test hash function that shards assuming -80 and 80- shards.

--- a/pkg/testutils/testing.go
+++ b/pkg/testutils/testing.go
@@ -13,6 +13,7 @@ import (
 	"testing"
 	"time"
 
+	"github.com/block/spirit/pkg/utils"
 	"github.com/go-sql-driver/mysql"
 	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/require"
@@ -153,15 +154,14 @@ func IsMinimalRBRTestRunner(t *testing.T) bool {
 }
 
 // WaitForReplicaHealthy polls SHOW REPLICA STATUS until both the IO and SQL
-// threads report Yes, or the timeout elapses. On timeout it skips the test
-// with an infra-attribution message — when the replica isn't running at the
-// start of a test, that's almost always a CI setup issue, not a Spirit bug,
-// so failing late inside m.Run() is misleading.
+// threads report Yes, or the timeout elapses. On timeout it fails the test
+// with an infra-attribution message so a broken CI replica setup is clearly
+// distinguishable from a Spirit migration bug.
 func WaitForReplicaHealthy(t *testing.T, dsn string, timeout time.Duration) {
 	t.Helper()
 	db, err := sql.Open("mysql", dsn)
 	require.NoError(t, err)
-	defer func() { _ = db.Close() }()
+	defer utils.CloseAndLog(db)
 
 	deadline := time.Now().Add(timeout)
 	var lastIO, lastSQL, lastIOState string
@@ -174,7 +174,7 @@ func WaitForReplicaHealthy(t *testing.T, dsn string, timeout time.Duration) {
 			lastIO, lastSQL, lastIOState = ioRunning, sqlRunning, ioState
 		}
 		if time.Now().After(deadline) {
-			t.Skipf("test infra: replica not healthy after %s "+
+			t.Fatalf("test infra: replica not healthy after %s "+
 				"(Replica_IO_Running=%q Replica_SQL_Running=%q Replica_IO_State=%q lastErr=%v); "+
 				"this indicates a CI setup issue, not a Spirit bug",
 				timeout, lastIO, lastSQL, lastIOState, err)
@@ -188,7 +188,7 @@ func readReplicaStatus(ctx context.Context, db *sql.DB) (ioRunning, sqlRunning, 
 	if err != nil {
 		return "", "", "", err
 	}
-	defer func() { _ = rows.Close() }()
+	defer utils.CloseAndLog(rows)
 	cols, err := rows.Columns()
 	if err != nil {
 		return "", "", "", err

